cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/FileSystemStorageAdapter.java

```diff
@@ -0,0 +1,523 @@
+                case URI_BASED:
+                    iterator = new ArtifactIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                case URI_BUCKET_BASED:
+                    iterator = new FileSystemIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                default:
+                    throw new IllegalStateException("unsupported bucket mode");
+            }
+            return iterator;
+        } catch (InvalidPathException e) {
+            throw new IllegalArgumentException("Invalid bucket: " + bucket);
+        }
+    }
+    
+    private Path createArtifactPath(URI artifactURI, String uriBucket) {
+        StringBuilder path = new StringBuilder();
+        switch (bucketMode) {
+            case URI_BASED:
+                path.append(artifactURI.getScheme());
+                path.append(":").append(File.separator);
+                path.append(artifactURI.getSchemeSpecificPart());
+                break;
+            case URI_BUCKET_BASED:
+                for (char c : uriBucket.toCharArray()) {
+                    path.append(c).append(File.separator);
+                }
+                UUID filename = UUID.randomUUID();
+                path.append(filename);
+                break;
+            default:
+                throw new IllegalStateException("unsupported bucket mode");
+        }
+        return root.resolve(path.toString());
+    }
+    
+    private Path createStorageLocationPath(StorageLocation storageLocation) throws ResourceNotFoundException {
+        URI storageID = storageLocation.getStorageID();
+        if (!storageID.getScheme().equals(STORAGE_URI_SCHEME)) {
+            throw new IllegalArgumentException("Unkown storage id scheme: " + storageID.getScheme());
+        }
+        StringBuilder path = new StringBuilder();
+        switch (bucketMode) {
+            case URI_BASED:
+                path.append(storageID.getSchemeSpecificPart());
+                break;
+            case URI_BUCKET_BASED:
+                String bucket = storageLocation.storageBucket;
+                for (char c : bucket.toCharArray()) {
+                    path.append(c).append(File.separator);
+                }
+                path.append(storageLocation.getStorageID().getSchemeSpecificPart());
+                break;
+            default:
+                throw new IllegalStateException("unsupported bucket mode");
+        }
+        
+        log.debug("Resolving path: " + path.toString());
+        Path ret = root.resolve(path.toString());
+        if (!Files.exists(ret)) {
+            throw new ResourceNotFoundException("not found: " + storageID);
+        }
+        if (!Files.isRegularFile(ret)) {
+            throw new IllegalArgumentException("not found: " + storageID);
+        }
+        return ret;
+    }
+    
+    public static URI createMD5Checksum(Path path) throws NoSuchAlgorithmException, IOException {
```

Comment:
What is the purpose of this method?

Reply:
It's used by the file system iterator.  I'm not sure why it's in this class though.  I'll move it.